service/topology_coordinator.cc

```diff
@@ -664,6 +665,119 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+        using host_idx_to_load_t = std::pair<size_t, double>;
+        std::vector<host_idx_to_load_t> host_load;
+        host_load.reserve(replicas.size());
+
+        for (size_t i = 0; i < replicas.size(); i++) {
+            host_idx_to_load_t host_load_pair;
+            if (host_id_to_dc(replicas[i].host) == dc) {
+                host_load_pair.first = i;
+                host_load_pair.second = load.get_load(replicas[i].host);
+                host_load.push_back(host_load_pair);
+            }
+        }
+
+        auto cmp_by_load = [](const auto& a, const auto& b) {
+            return a.second > b.second;
+        };
+        auto cmp_by_idx = [](const auto& a, const auto& b) {
+            return a.first > b.first;
+        };
+
+        size_t to_remove = current_rf - new_rf;
+
+        // ensure that the most loaded replicas are in the beginning
+        std::nth_element(host_load.begin(), host_load.begin() + to_remove,
+            host_load.end(), cmp_by_load);
+
+        // sort the most loaded replicas by id, descending so that
+        // we can remove them by their index
+        std::sort(host_load.begin(), host_load.begin() + to_remove, cmp_by_idx);
+
+        for (size_t i = 0; i < to_remove; i++) {
+            replicas.erase(replicas.begin() + host_load[i].first);
+        }
+    }
+
+    static void upsize_tablet_replicas_to_rf(locator::tablet_replica_set& replicas, size_t current_rf,
+        size_t new_rf, locator::load_sketch& load, const locator::token_metadata_ptr& tm, const sstring& dc)
+    {
+        for (const auto& endpoint : tm->get_topology().get_datacenter_endpoints().at(dc)) {
+            locator::host_id ep = tm->get_topology().find_node(endpoint)->host_id();
+            replicas.emplace_back(locator::tablet_replica{ep, load.next_shard(ep)});
+            current_rf++;
+            if (current_rf == new_rf)
+                break;
+        }
+    }
+
+    future<locator::tablet_map>
+    reallocate_tablets_for_new_rf(schema_ptr s, locator::token_metadata_ptr tm,
+        std::unordered_map<sstring, size_t> new_dc_rep_factor) const {
+        // FIXME: Don't use tokens to distribute nodes.
+        // The following reuses the existing token-based algorithm used by NetworkTopologyStrategy.
+        assert(!tm->sorted_tokens().empty());
+
+        auto table_id = s->id();
+        locator::tablet_map tablets = tm->tablets().get_tablet_map(table_id);
+
+        locator::load_sketch load(tm);
+        co_await load.populate();
+
+        auto host_id_to_dc = [&tm](const locator::host_id& ep) -> std::optional<sstring> {
+            auto node = tm->get_topology().find_node(ep);
+            if (node == nullptr) {
+                return std::nullopt;
+            }
+            return node->dc_rack().dc;
+        };
+
+        for (auto tb : tablets.tablet_ids()) {
+            const locator::tablet_info& ti = tablets.get_tablet_info(tb);
+            std::unordered_map<sstring, size_t> dc_rep_factor;
```

Comment:
If this is the old replication map, then it either has to be extracted from query_processor using the schema_ptr or passed to this function as one of its params.

Reply:
This is the old replication map, is retrieving it through the token metadata inappropriate? Token metadata is necessary in the function anyway, for `load_sketch`.